File: app/src/main/java/com/tmvlg/smsreceiver/backend/FreeNumbersParser.java
package com.tmvlg.smsreceiver.backend;

import android.os.AsyncTask;
import android.util.Log;

import com.google.i18n.phonenumbers.NumberParseException;
import com.google.i18n.phonenumbers.PhoneNumberUtil;
import com.google.i18n.phonenumbers.Phonenumber;
import com.neovisionaries.i18n.CountryCode;

import org.hamcrest.core.Is;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.JSONValue;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Locale;
import java.util.Scanner;

public class FreeNumbersParser {

    private String TAG = "1";

    public ArrayList<FreeNumber> numbersList;

    public ArrayList<FreeMessage> messagesList;





    /////////////////////////////////////////////////////////////
    //////////https://sms-online.co/receive-free-sms/////////////
    /////////////////////////////////////////////////////////////


    public void parse_1() {
        try {
            Document doc = Jsoup.connect("https://sms-online.co/receive-free-sms").get();
            Elements numbers = doc.select("h4.number-boxes-item-number");
            Elements countries = doc.select("h5.number-boxes-item-country");

            Log.d(TAG, numbers.text());

//            CountryCodes codes = new CountryCodes();
            for (int i = 0; i < numbers.size(); i++) {
//                String code = codes.getCode(countries.get(i).text());
                Log.d(TAG, "parse_1: country_name = " + countries.get(i).text());
                String code =  CountryCode.findByName(countries.get(i).text()).get(0).name();


                if (code == null) {
                    Log.d(TAG, "null country code at " + countries.get(i).text());
                    continue;
                }
//                String prefix = Iso2Phone.getPhone(code);

                int prefix2 = PhoneNumberUtil.getInstance().getCountryCodeForRegion(code);
                String prefix = "+" + prefix2;
                Log.d(TAG, "parse_1: prefix2 = " + prefix2);

                if (prefix == null || prefix.contains("and")) {
                    Log.d(TAG, "null number prefix at " + code);
                    continue;
                }

                FreeNumber free_number = new FreeNumber();

                free_number.call_number = numbers.get(i).text().replace(prefix, "").replaceAll("^\\s+", "");
                free_number.call_number_prefix = prefix;
                free_number.country_code = code;
                free_number.counrty_name = countries.get(i).text();
                free_number.origin = "parse_1";

                numbersList.add(free_number);
            }
        } catch (IOException e) {
            Log.d(TAG, "parsing error at parse_1");
            e.printStackTrace();
        }

    }



    public void parse_messages_1(String callnumber) {
        try {
            Document doc = Jsoup.connect("https://sms-online.co/receive-free-sms/" + callnumber).get();
            Elements titles = doc.select("h3.list-item-title");
            Elements messages = doc.select("div.list-item-content");
            Elements times = doc.select("span.list-item-meta > span");

            for (int i = 0; i < titles.size(); i++) {
                FreeMessage free_message = new FreeMessage();

                free_message.message_title = titles.get(i).text();
                free_message.message_text = messages.get(i).text();
                free_message.time_remained = times.get(i).text();

                messagesList.add(free_message);
            }


        } catch (IOException e) {
            Log.d(TAG, "parsing error at parse_messages_1");
            e.printStackTrace();
        }
    }


    /////////////////////////////////////////////////////////////
    ////////////////////https:///smska.us////////////////////////
    /////////////////////////////////////////////////////////////


    public void parse_2() {
        try {
            Document doc = Jsoup.connect("https://smska.us/").get();
            Elements numbers = doc.select("div.phone_number");

            for (int i = 0; i < numbers.size(); i++) {

                FreeNumber free_number = new FreeNumber();

                String prefix = "+7";
                String code = "RU";

                free_number.call_number = numbers.get(i).text().replace(prefix, "").replaceAll("^\\s+", "");
                free_number.call_number_prefix = prefix;
                free_number.country_code = code;
                free_number.counrty_name = "Russia";
                free_number.origin = "parse_2";

                numbersList.add(free_number);
            }
        } catch (IOException e) {
            Log.d(TAG, "parsing error at parse_2");
            e.printStackTrace();
        }

    }


    public void parse_messages_2(String callnumber) {
        try {
            Document doc = Jsoup.connect("https://smska.us/").get();
            Elements titles = doc.select("div#" + callnumber + " div:nth-child(1)");
            Elements messages = doc.select("div#" + callnumber + " > div.textsms");
            Elements times = doc.select("div#" + callnumber + " > div.smsdate");

            Log.d(TAG, "div#" + callnumber + " > div.bodysms");
            Log.d(TAG, titles.html());


            for (int i = 0; i < titles.size(); i++) {
                Log.d(TAG, "iteration++");
                Log.d(TAG, titles.get(i).text());

                FreeMessage free_message = new FreeMessage();

                free_message.message_title = titles.get(i).text();
                free_message.message_text = messages.get(i).text();
                free_message.time_remained = times.get(i).text();

                messagesList.add(free_message);
            }


        } catch (IOException e) {
            Log.d(TAG, "parsing error at parse_messages_2");
            e.printStackTrace();
        }
    }




    /////////////////////////////////////////////////////////////
    ////////////////////https://online-sms.org/////////////////
    /////////////////////////////////////////////////////////////


    public void parse_3() {
        try {
            Document doc = Jsoup.connect("https://online-sms.org/").get();
            Elements countries = doc.select(("h4.titlecoutry"));



            CountryCodes codes = new CountryCodes();
            for (int i = 0; i < countries.size(); i++) {
                Log.d(TAG, "parse_3: country = " + countries.get(i).text());
                String code = codes.getCode(countries.get(i).text());
//                String code = convertCountryNameToIsoCode(countries.get(i).text()).getDisplayCountry();
                Log.d(TAG, "parse_3: code = " + code);

                if (code == null) {
                    Log.d(TAG, "null country code at " + countries.get(i).text());
                    try {
                        code = CountryCode.findByName(countries.get(i).text()).get(0).name();
                    }
                    catch (Exception e){
                        Log.d(TAG, "parse_3: null cc again");
                        continue;
                    }
                }
//                String prefix = Iso2Phone.getPhone(code);
                int prefix2 = PhoneNumberUtil.getInstance().getCountryCodeForRegion(code);
                String prefix = "+" + prefix2;
                Log.d(TAG, "parse_3: prefix = " + prefix);
                if (prefix == null || prefix.contains("and")) {
                    Log.d(TAG, "null number prefix at " + code);
                    continue;
                }

                Elements numbers = doc.select("#tab" + code + " a");
                Log.d(TAG, "parse_3: elemnumbs = " + numbers.text());
                Log.d(TAG, "#tab" + code + " a");

                for (int j = 0; j < numbers.size(); j++) {
                    Log.d(TAG, numbers.get(j).text());

                    FreeNumber free_number = new FreeNumber();

                    free_number.call_number = numbers.get(j).text().replace(prefix, "").replaceAll("^\\s+", "");
                    free_number.call_number_prefix = prefix;
                    free_number.country_code = code;
                    free_number.counrty_name = countries.get(i).text();
                    free_number.origin = "parse_3";

                    numbersList.add(free_number);
                }

            }
        } catch (IOException e) {
            Log.d(TAG, "parsing error at parse_3");
            e.printStackTrace();
        }

    }


    public void parse_messages_3(String callnumber) {
        try {
            Document doc = Jsoup.connect("https://online-sms.org/free-phone-number-" + callnumber).get();
            Elements titles = doc.select("table > tbody > tr > td:nth-child(1)");
            Elements messages = doc.select("table > tbody > tr > td:nth-child(2)");
            Elements times = doc.select("table > tbody > tr > td:nth-child(3)");

            Log.d(TAG, "div#" + callnumber + " > div.bodysms");
            Log.d(TAG, titles.html());

            for (int i = 0; i < titles.size(); i++) {
                Log.d(TAG, "iteration++");
                Log.d(TAG, titles.get(i).text());

                FreeMessage free_message = new FreeMessage();

                free_message.message_title = titles.get(i).text();
                free_message.message_text = messages.get(i).text();
                free_message.time_remained = times.get(i).text();

                messagesList.add(free_message);
            }


        } catch (IOException e) {
            Log.d(TAG, "parsing error at parse_messages_3");
            e.printStackTrace();
        }
    }

    public void parse_4() throws ParseException, MalformedURLException {
        String inline_countries = parse_inline(new URL("https://onlinesim.ru/api/getFreeCountryList"));
        JSONParser parse_countries = new JSONParser();
        JSONObject jobj_countries = (JSONObject)parse_countries.parse(inline_countries);
        JSONArray countries_Array = (JSONArray) jobj_countries.get("countries");
        for (int i = 0; i < countries_Array.size(); i++) {
            JSONObject jsonCountry = (JSONObject)countries_Array.get(i);
            System.out.print(" << jsonCountry >> ");
            System.out.println(jsonCountry);
            // int local_country_code = Integer.parseInt((String)jsonCountry.get("country"));
            String country_name_local = (String)jsonCountry.get("country_text");
            int local_country_code = (int) (long) jsonCountry.get("country");


            String inline_numbers = parse_inline(new URL("https://onlinesim.ru/api/getFreePhoneList?country=" + local_country_code));
            JSONParser parse_numbers = new JSONParser();
            JSONObject jobj_numbers = (JSONObject)parse_numbers.parse(inline_numbers);
            JSONArray numbers_Array = (JSONArray) jobj_numbers.get("numbers");
            for (int j = 0; j < numbers_Array.size(); j++) {
                JSONObject jsonNumber = (JSONObject) numbers_Array.get(j);
                System.out.print(" >>>> jsonNumber: ");
                System.out.println(jsonNumber);
                String call_number = (String) jsonNumber.get("number");
                String full_number = (String) jsonNumber.get("full_number");
                String country_prefix = full_number.replace(call_number, "");

                PhoneNumberUtil phoneUtil = PhoneNumberUtil.getInstance();

                String country_code = "";
                try {
                    // phone must begin with '+'
                    Phonenumber.PhoneNumber numberProto = phoneUtil.parse(full_number, "");
                    country_code = phoneUtil.getRegionCodeForCountryCode(numberProto.getCountryCode());
                } catch (Exception e) {
                    System.err.println("NumberParseException was thrown: " + e.toString());
                }

                Locale locale = new Locale("", country_code);
                String country_name = locale.getDisplayCountry();
                Log.d(TAG, "parse_4: country_name " + country_name);

                FreeNumber temp = new FreeNumber();
//                temp.local_country = country_name;
//                temp.local_country_code = local_country_code;
                temp.call_number = call_number;
                temp.country_code = country_code;
                temp.country_code = country_code;
                temp.counrty_name = country_name;
                temp.call_number_prefix = country_prefix;
                temp.origin = "parse_4";

                numbersList.add(temp);


            }
        }
    }




    public void parse_messages_4(String phone) throws ParseException, MalformedURLException {
        String inline_messages = parse_inline(new URL("https://onlinesim.ru/api/getFreeMessageList?cpage=1&phone=" + phone));
        Log.d(TAG, "parse_messages_4: inline = " + inline_messages);
        JSONParser parse_messages = new JSONParser();
        JSONObject jobj_messages = (JSONObject)parse_messages.parse(inline_messages);
        JSONArray messages_Array = (JSONArray) jobj_messages.get("messages");
        Log.d(TAG, "parse_messages_4: MessagesAray size: " + messages_Array.size());
        for (int i = 0; i < messages_Array.size(); i++) {
            JSONObject jsonMessage = (JSONObject)messages_Array.get(i);
            System.out.print(" << jsonMessage >> ");
            System.out.println(jsonMessage);
//            String country_name_local = (String)jsonMessage.get("country_text");
//            int local_country_code = (int) (long) jsonMessage.get("country");
        }
    }



    public void parse_numbers() throws ParseException, MalformedURLException {
//        parse_1();
//        parse_2();
        parse_3();
//        parse_4();

        sort_numbers();
    }

    public void parse_messages(String origin, String callnumber) throws MalformedURLException, ParseException {
        messagesList.clear();
        if (origin.equals("parse_1")) {
            Log.d(TAG, callnumber);
            parse_messages_1(callnumber);
        }
        else if (origin.equals("parse_2")) {
            Log.d(TAG, callnumber.substring(1));
            parse_messages_2(callnumber.substring(1));
        }
        else if (origin.equals("parse_3")) {
            Log.d(TAG, callnumber.substring(1));
            parse_messages_3(callnumber);
        }
        else if (origin.equals("parse_4")) {
            Log.d(TAG, callnumber);
            parse_messages_4("+" + callnumber);
        }
    }


    public static String parse_inline(URL url) {
        // URL url = new URL("https://google.com");

        try {

            HttpURLConnection conn = (HttpURLConnection)url.openConnection();
            conn.setRequestMethod("GET");
            conn.connect();
            int responsecode = conn.getResponseCode();
            String inline = "";

            if(responsecode != 200)
                throw new RuntimeException("HttpResponseCode: " +responsecode);
            else
            {
                Scanner sc = new Scanner(url.openStream());
                while(sc.hasNext())
                {
                    inline+=sc.nextLine();
                }
                System.out.println("INLINE FETCHED!");
                sc.close();
                return inline;
            }
        }
        catch (Exception e) {
            System.out.println(e.getMessage());
        }
        return "";
    }


    public void sort_numbers() {
        Collections.sort(numbersList, new Comparator<FreeNumber>() {
            @Override
            public int compare(FreeNumber o1, FreeNumber o2) {
                int comparision = o1.counrty_name.compareTo(o2.counrty_name);
                return Integer.compare(comparision, 0);
            }
        });
    }

    public void print_data() {
        for (FreeNumber cur_number : numbersList) {
            Log.d(TAG, cur_number.call_number_prefix + " : " + cur_number.call_number + " : " + cur_number.country_code);
        }
    }

    Locale convertCountryNameToIsoCode(String countryName){
        for(Locale l : Locale.getAvailableLocales()) {
            if (l.getDisplayCountry().equals(countryName)) {
                return l;
            }
        }
        return null;
    }

    public FreeNumbersParser () {
        numbersList = new ArrayList<>();
        messagesList = new ArrayList<>();
    }



}

File: app/src/main/java/com/tmvlg/smsreceiver/data/FreeNumbersParser.kt
package com.tmvlg.smsreceiver.data

import android.util.Log
import com.google.i18n.phonenumbers.PhoneNumberUtil
import com.tmvlg.smsreceiver.data.network.freenumbersapi.onlinesimfree.OnlineSimFreeApi
import com.tmvlg.smsreceiver.data.network.freenumbersapi.onlinesimfree.models.FreeCountriesResponse
import com.tmvlg.smsreceiver.data.network.freenumbersapi.onlinesimfree.models.MessagesResponse
import com.tmvlg.smsreceiver.data.network.freenumbersapi.onlinesimfree.models.NumbersResponse
import com.tmvlg.smsreceiver.domain.freemessage.FreeMessage
import com.tmvlg.smsreceiver.domain.freenumber.FreeNumber
import com.tmvlg.smsreceiver.util.*
import org.json.simple.parser.ParseException
import org.jsoup.Jsoup
import retrofit2.Response
import java.io.IOException
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.SocketTimeoutException
import java.net.URL
import java.text.SimpleDateFormat
import java.util.*

class FreeNumbersParser {

    /////////////////////////////////////////////////////////////
    //////////https://sms-online.co/receive-free-sms/////////////
    /////////////////////////////////////////////////////////////

    fun parse_1(): List<FreeNumber> {
        val numbersList = mutableListOf<FreeNumber>()
        try {
            val doc = Jsoup.connect("https://sms-online.co/receive-free-sms").get()
            val numbers = doc.select("h4.number-boxes-item-number")
            val countries = doc.select("h5.number-boxes-item-country")
            Log.d(TAG, numbers.text())
            for (i in numbers.indices) {
                val code = getCountryCodeByName(countries[i].text()) ?: continue
                val prefix = getPrefixByCountryCode(code) ?: continue
                val countryName = countries[i].text()
                val free_number = FreeNumber(
                    call_number = numbers[i].text().replace(prefix, "")
                        .replace("^\\s+".toRegex(), ""),
                    call_number_prefix = prefix,
                    country_code = code,
                    counrty_name = code.translateCountryName(),
                    origin = "parse_1",
                    icon_path = "$code.xml"
                )
                numbersList.add(free_number)
            }
        } catch (t: Throwable) {
            Log.d(TAG, "parsing error at parse_1")
            t.printStackTrace()
        }
        return numbersList
    }

    fun parse_messages_1(callnumber: String): List<FreeMessage> {
        val messagesList = mutableListOf<FreeMessage>()
        try {
            val doc = Jsoup.connect("https://sms-online.co/receive-free-sms/$callnumber").get()
            val titles = doc.select("h3.list-item-title")
            val messages = doc.select("div.list-item-content")
            val times = doc.select("span.list-item-meta > span")
            for (i in titles.indices) {
                val free_message = FreeMessage(
                    message_title = titles[i].text(),
                    message_text = messages[i].text(),
                    time_remained = times[i].text()
                )
                messagesList.add(free_message)
            }
        } catch (t: Throwable) {
            Log.d(TAG, "parsing error at parse_messages_1")
            t.printStackTrace()
        }
        return messagesList
    }

    /////////////////////////////////////////////////////////////
    ////////////////////https:///smska.us////////////////////////
    /////////////////////////////////////////////////////////////
    fun parse_2(): List<FreeNumber> {
        val numbersList = mutableListOf<FreeNumber>()
        try {
            val doc = Jsoup.connect("https://smska.us/").get()
            val numbers = doc.select("div.phone_number")
            for (i in numbers.indices) {
                val prefix = "+7"
                val code = "RU"
                val countryName = "Russia"
                val free_number = FreeNumber(
                    call_number = numbers[i].text().replace(prefix, "")
                        .replace("^\\s+".toRegex(), ""),
                    call_number_prefix = prefix,
                    country_code = code,
                    counrty_name = code.translateCountryName(),
                    origin = "parse_2",
                    icon_path = "$code.xml"
                )
                numbersList.add(free_number)
            }
        } catch (t: Throwable) {
            Log.d(TAG, "parsing error at parse_2")
            t.printStackTrace()
        }
        return numbersList
    }

    fun parse_messages_2(callnumber: String): List<FreeMessage> {
        val messagesList = mutableListOf<FreeMessage>()
        try {
            val doc = Jsoup.connect("https://smska.us/AjaxPublic/sms/")
                .header("Content-Type", "application/x-www-form-urlencoded")
                .header("X-Requested-With", "XMLHttpRequest")
                .data("n", callnumber)
                .post()

            val titles = doc.select("div.bodysms > div.smsnumber")
            val messages = doc.select("div.bodysms > div.textsms")
            val times = doc.select("div.bodysms > div.smsdate")

            for (i in titles.indices) {
                val time = times[i].text()
                val dateTime = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
                    .parse(time)
                    ?: return messagesList
                val timeFormatted = Date(dateTime.time).getTimeAgo(TimeZone.getTimeZone("Europe/Moscow"))
                val free_message = FreeMessage(
                    message_title = titles[i].text(),
                    message_text = messages[i].text(),
                    time_remained = timeFormatted
                )
                messagesList.add(free_message)
            }

        } catch (t: Throwable) {
            Log.d(TAG, "parsing error at parse_messages_2")
            t.printStackTrace()
        }
        return messagesList
    }

    /////////////////////////////////////////////////////////////
    ////////////////////https://online-sms.org/////////////////
    /////////////////////////////////////////////////////////////
    fun parse_3(): List<FreeNumber> {
        val numbersList = mutableListOf<FreeNumber>()
        try {
            val doc = Jsoup.connect("https://online-sms.org/").get()
            val countries = doc.select("h4.titlecoutry")
            for (i in countries.indices) {
                val countryName = countries[i].text()
                Log.d(TAG, countryName)
                val code = getCountryCodeByName(countryName) ?: continue
                val prefix = getPrefixByCountryCode(code)
                if (prefix == null || prefix.contains("and")) {
                    Log.d(TAG, "null number prefix at $code")
                    continue
                }
                val numbers = doc.select("#tab$code a")
                Log.d(TAG, "#tab$code a")
                for (j in numbers.indices) {
                    val number = numbers[j].text()
                    Log.d(TAG, number)
                    val numberFormatted = numbers[j].text().replace(prefix, "")
                        .replace("^\\s+".toRegex(), "")
                    val free_number = FreeNumber(
                        call_number = numberFormatted,
                        call_number_prefix = prefix,
                        country_code = code,
                        counrty_name = code.translateCountryName(),
                        origin = "parse_3",
                        icon_path = "$code.xml"
                    )
                    numbersList.add(free_number)
                }
            }
        } catch (t: Throwable) {
            Log.d(TAG, "parsing error at parse_3")
            t.printStackTrace()
        }
        return numbersList
    }

    fun parse_messages_3(callnumber: String): List<FreeMessage> {
        val messagesList = mutableListOf<FreeMessage>()
        try {
            val doc = Jsoup.connect("https://online-sms.org/free-phone-number-$callnumber").get()
            val titles = doc.select("table > tbody > tr > td:nth-child(1)")
            val messages = doc.select("table > tbody > tr > td:nth-child(2)")
            val times = doc.select("table > tbody > tr > td:nth-child(3)")
            Log.d(TAG, "div#$callnumber > div.bodysms")
            Log.d(TAG, titles.html())
            for (i in titles.indices) {
                Log.d(TAG, "iteration++")
                Log.d(TAG, titles[i].text())
                val free_message = FreeMessage(
                    message_title = titles[i].text(),
                    message_text = messages[i].text(),
                    time_remained = times[i].text()
                )
                messagesList.add(free_message)
            }
        } catch (t: Throwable) {
            Log.d(TAG, "parsing error at parse_messages_3")
            t.printStackTrace()
        }
        return messagesList
    }

    fun parse_4(): List<FreeNumber> {
        val numbersList = mutableListOf<FreeNumber>()
        try {
            val freeCountriesResponse: Response<FreeCountriesResponse> = OnlineSimFreeApi
                .retrofitService
                .getCountryNames()
                .execute() ?: return numbersList
            val countriesBody = freeCountriesResponse.body() ?: return numbersList
            countriesBody.countries.forEach {

                val numbersResponse: Response<NumbersResponse> = OnlineSimFreeApi
                    .retrofitService
                    .getFreeNumbers(it.country)
                    .execute()

                val numbersBody: NumbersResponse = numbersResponse.body() ?: return numbersList
                numbersBody.numbers.forEach {

                    val countryCode: String = getCountryCodeByNumber(it.fullNumber)
                    val countryName = getCountryNameByCode(countryCode) ?: return numbersList

                    val freeNumber = FreeNumber(
                        call_number = it.number,
                        call_number_prefix = "+${it.country}",
                        country_code = countryCode,
                        counrty_name = countryCode.translateCountryName(),
                        origin = "parse_4",
                        icon_path = "${countryCode.uppercase()}.xml"
                    )
                    numbersList.add(freeNumber)
                }
            }
        } catch (t: Throwable) {
            if (t is SocketTimeoutException)
                Log.d(TAG, "parse_messages_4: timeout")
            Log.d(TAG, "parse_messages_4 error")
        }
        return numbersList
    }

    fun parse_messages_4(phone: String, prefix: String): List<FreeMessage> {
        val messagesList = mutableListOf<FreeMessage>()
        try {
            val messagesResponse: Response<MessagesResponse> = OnlineSimFreeApi
                .retrofitService
                .getFreeMessages(phone, prefix)
                .execute() ?: return messagesList
            val messagesBody = messagesResponse.body() ?: return messagesList
            messagesBody.messages.data.forEach {
                val free_message = FreeMessage(
                    message_title = it.inNumber,
                    message_text = it.text,
                    time_remained = it.dataHumans
                )
                messagesList.add(free_message)

            }
        } catch (t: Throwable) {
            if (t is SocketTimeoutException)
                Log.d(TAG, "parse_messages_4: timeout")
            Log.d(TAG, "parse_messages_4 error")
        }
        return messagesList
    }

    fun parse_5(): List<FreeNumber> {
        val numbersList = mutableListOf<FreeNumber>()
        try {
            val doc = Jsoup.connect("https://receive-smss.com/").get()
            val numbers = doc.select("h4.number-boxes-itemm-number")
            val countries = doc.select("h5.number-boxes-item-country")
            val numberType = doc.select("a.number-boxes-item-button")

            for (i in numbers.indices) {
                if (numberType[i].text() == "premium") {
                    continue
                } else {
                    val code = getCountryCodeByName(countries[i].text()) ?: continue
                    val prefix = getPrefixByCountryCode(code) ?: continue
                    val countryName = countries[i].text()
                    val free_number = FreeNumber(
                        call_number = numbers[i].text().replace(prefix, "")
                            .replace("^\\s+".toRegex(), ""),
                        call_number_prefix = prefix,
                        country_code = code,
                        counrty_name = code.translateCountryName(),
                        origin = "parse_5",
                        icon_path = "$code.xml"
                    )
                    numbersList.add(free_number)
                }
            }

        } catch (t: Throwable) {
            t.printStackTrace()
        }
        return numbersList
    }

    fun parse_messages_5(number: String): List<FreeMessage> {
        val messagesList = mutableListOf<FreeMessage>()
        try {
            val doc = Jsoup.connect("https://receive-smss.com/sms/$number/").get()

            val sender = doc.select("tr > td.wr3pc52sel1757:nth-child(1)")
//            val messageCode = doc.select("span.btncp1 > b")
            val messageBody = doc.select("tr > td.wr3pc52sel1757:nth-child(2)")
            val time = doc.select("tr > td.wr3pc52sel1757:nth-child(3) > span")

            for (i in sender.indices) {
                val messageTitleFormatted = sender[i].text().replace("a\"", "")
                val timeRemainedFormatted = "${time[i].text()} ago"
                val free_message = FreeMessage(
                    message_title = messageTitleFormatted,
                    message_text = messageBody[i].text(),
                    time_remained = timeRemainedFormatted
                )
                messagesList.add(free_message)
            }
            println(messagesList)

        } catch (t: Throwable) {
            println("parsing error")
            t.printStackTrace()
        }
        return messagesList
    }

    fun parse_numbers(): List<FreeNumber> {
        val tempFreeNumberList = mutableListOf<FreeNumber>()
        tempFreeNumberList += parse_1()
        tempFreeNumberList += parse_2()
        tempFreeNumberList += parse_3()
        tempFreeNumberList += parse_4()
        tempFreeNumberList += parse_5()
        add_countrynames_to_list(tempFreeNumberList)
        sort_numbers(tempFreeNumberList)
//        for (number in tempFreeNumberList) {
//            Log.d(TAG, "parse_numbers: $number")
//        }
        Log.d(TAG, "parse_numbers_size: ${tempFreeNumberList.size}")
        return tempFreeNumberList
    }

    fun parse_messages(freeNumber: FreeNumber): List<FreeMessage> {
        Log.d(TAG, "parse_messages: STARTED")
        val tempMessagesList = mutableListOf<FreeMessage>()
        with(freeNumber) {
            var formated_call_number = call_number
            formated_call_number = call_number_prefix
                .substring(1) + formated_call_number
                .replace(" ", "")
                .replace("-", "")
            when (origin) {
                "parse_1" -> {
                    Log.d(TAG, formated_call_number)
                    tempMessagesList += parse_messages_1(formated_call_number)
                }
                "parse_2" -> {
                    Log.d(TAG, formated_call_number.substring(1))
                    tempMessagesList += parse_messages_2(formated_call_number.substring(1))
                }
                "parse_3" -> {
//                    Log.d(TAG, call_number.substring(1))
                    Log.d(TAG, "parse_messages: $formated_call_number")
                    tempMessagesList += parse_messages_3(formated_call_number)
                }
                "parse_4" -> {
                    Log.d(TAG, call_number)
                    tempMessagesList += parse_messages_4(call_number, call_number_prefix)
                }
                "parse_5" -> {
                    Log.d(TAG, call_number)
                    tempMessagesList += parse_messages_5(formated_call_number)
                }
                else -> throw java.lang.RuntimeException("Unknown parse origin")
            }
        }
        return tempMessagesList
    }

    fun sort_numbers(numbersList: List<FreeNumber>) {
        (numbersList as MutableList<FreeNumber>).sortWith { o1, o2 ->
            val comparision = o1.counrty_name.compareTo(o2.counrty_name)
            comparision.compareTo(0)
        }
    }

    fun add_countrynames_to_list(numbersList: List<FreeNumber>) {
        if (numbersList.isEmpty())
            return
        var countryName = ""
        val tempNumberList = mutableListOf<FreeNumber>()
        for (item in numbersList) {
            val new_countryName = item.counrty_name
            val new_call_number_prefix = item.call_number_prefix
            val new_country_code = item.country_code
            if (new_countryName != countryName) {
                val free_number = FreeNumber(
                    call_number = "",
                    call_number_prefix = new_call_number_prefix,
                    country_code = new_country_code,
                    counrty_name = new_countryName,
                    origin = "",
                    icon_path = "",
                    type = FreeNumber.HEADER_TYPE
                )
                if (!tempNumberList.containsCountryName(free_number.counrty_name)) {
                    tempNumberList.add(free_number)
                    countryName = new_countryName
                }
            }
        }
        for (item in tempNumberList) {
            (numbersList as MutableList<FreeNumber>).add(0, item)
        }
    }

    fun print_data(numbersList: List<FreeNumber>) {
        for (cur_number in numbersList) {
            Log.d(
                TAG,
                cur_number.call_number_prefix + " : " + cur_number.call_number + " : " + cur_number.country_code
            )
        }
    }


    companion object {
        private const val TAG = "1"

        fun parse_inline(url: URL): String {
            // URL url = new URL("https://google.com");
            try {
                val conn = url.openConnection() as HttpURLConnection
                conn.requestMethod = "GET"
                conn.connect()
                val responsecode = conn.responseCode
                var inline = ""
                return if (responsecode != 200) throw RuntimeException("HttpResponseCode: $responsecode") else {
                    val sc = Scanner(url.openStream())
                    while (sc.hasNext()) {
                        inline += sc.nextLine()
                    }
                    println("INLINE FETCHED!")
                    sc.close()
                    inline
                }
            } catch (e: Exception) {
                println(e.message)
            }
            return ""
        }
    }
}